### DISCUSSION TITLE
Feature Request: partial reading for CSV files (smart column parsing)

### DISCUSSION THREAD

**Discussion by tbeason** [duckdb/duckdb#6337] [Ideas]
Why not make `read_csv` work more like reading from Parquet files, where it only parses and reads in the columns that it actually needs? I'm guessing that given the `SELECT` statement and the column headings in the file, the logic is mostly the same, isn't it?

Or, to go only part of the way there, perhaps the existing `columns=` argument to `read_csv` could behave like this. It would read only the columns specified.

This could greatly improve import performance on large CSV files.

**Comment by Mytherin** [comment]
Adding projection pushdown to the CSV reader is on our roadmap - but the performance gain is significantly smaller for CSV files than for Parquet files. For CSV files you will still need to parse all of the delimiters and read all of the bytes of the file. I have benchmarked this before and the performance gains are definitely there, but even if you select only 1 column out of 20, the reading will only become 2-3x faster, not 20x as with the Parquet reader.

**Reply by tbeason** [reply]
Glad to hear. I think it would be very worthwhile!
